src/main/java/edu/njit/jerse/automation/AsheAutomation.java

```diff
@@ -0,0 +1,211 @@
+        }
+
+        CompilationUnit cu = StaticJavaParser.parse(javaFilePath);
+
+        // targetFile - the Java file ASHE will target for minimization and error correction
+        // Example: edu/njit/jerse/automation/AsheAutomation.java
+        String targetFile = formatRelativePathForJavaFile(javaFilePath, projectRootPath);
+
+        Optional<String> packageName = cu.getPackageDeclaration()
+                .map(NodeWithName::getNameAsString);
+
+        // Example: "edu.njit.jerse.automation."
+        String packagePrefix = packageName.isEmpty() ? "" : packageName + ".";
+
+        for (TypeDeclaration<?> type : cu.getTypes()) {
+            if (type.isPublic()) {
+                // Example: AsheAutomation
+                String className = type.getNameAsString();
+
+                // Example: edu.njit.jerse.automation.AsheAutomation
+                String packageAndClassName = packagePrefix + className;
+
+                for (BodyDeclaration<?> member : type.getMembers()) {
+                    if (member instanceof MethodDeclaration method) {
+                        if (method.isPublic()) {
+                            // targetMethod - the method ASHE will target for minimization and error correction
+                            // Example: edu.njit.jerse.automation.AsheAutomation#main(String[])
+                            String targetMethod = fullyQualifiedMethodReference(packageAndClassName, method);
+
+                            ASHE ashe = new ASHE();
+                            ashe.run(projectRootPath, targetFile, targetMethod);
+                        }
+                    }
+                }
+            }
+        }
+        LOGGER.info("Completed processing Java file: {}", javaFileAbsolutePath);
+    }
+
+    /**
+     * Finds the relative path from the project root to the specified Java file. It formats the relative path
+     * to use forward slashes and strips out the leading source directory path.
+     *
+     * @param javaFilePath    the file path for which the relative path is needed
+     *                        Example: /absolute/path/src/main/java/com/example/foo/Bar.java
+     * @param projectRootPath the absolute path to the root of the project. This path
+     *                        must be a prefix of the javaFile's absolute path.
+     *                        Example: /absolute/path/src/main/java
+     * @return the relative path from the project root to the Java file
+     * Example of a relative path: com/example/foo/Bar.java
+     */
+    private static String formatRelativePathForJavaFile(Path javaFilePath, String projectRootPath) {
+        Path projectRoot = Paths.get(projectRootPath);
+        Path absoluteFilePath = javaFilePath.toAbsolutePath();
+
+        String relativePath = projectRoot.relativize(absoluteFilePath).toString();
+        relativePath = relativePath.replace(File.separatorChar, '/');
+
+        if (relativePath.startsWith(JAVA_SOURCE_DIR)) {
```

Comment:
This is a heuristic to make it easier to handle Gradle/Maven projects, right? If so, we should add a comment here that makes that explicit.

My bigger concern about this is projects with subprojects, though. A common structure for Gradle/Maven projects is that the top-level directory contains many subproject folders, each of which contains a `src/main/java` directory. Will this code still work under that scenario?

Reply:
Actually, it looks like my concern is probably handled by `ProjectRootFinder#findProjectRoots`. I'm still curious how this code will break if a project uses a non-standard directory structure, though.

Reply:
The above code should be able to find subprojects. There may be problems when ran against large repositories, but in general I think the development was intended for that. "I'm still curious how this code will break if a project uses a non-standard directory structure, though." This concern is what I have been thinking about over the past couple of weeks - what scenarios are there that do not follow Java standard directories? Will the use of [Java Modules](https://www.oracle.com/corporate/features/understanding-java-9-modules.html) hurt this in anyway?

The below image was found [here](https://www.baeldung.com/java-9-modularity):
<img width="523" alt="Screenshot 2023-11-27 at 8 16 35 PM" src="https://github.com/jonathan-m-phillips/ASHE_Automated-Software-Hardening-for-Entrypoints/assets/65977328/db642425-895b-475e-9533-452c10c1c2d9">

My thoughts are that what we have right now for the initial automation might be enough, and add more specific cases as we move forward. This can (and probably should) be an issue created, unless you or @mernst have other thoughts and would like to add it to this PR. I believe we should try to keep the simplicity by checking for standard Java dirs for the time being.

Reply:
And with Java Modules, I believe that `module-info.java` is required. Meaning we could check for that file within a dir to combat this.

Reply:
> Actually, it looks like my concern is probably handled by `ProjectRootFinder#findProjectRoots`. I'm still curious how this code will break if a project uses a non-standard directory structure, though.

Also, I do not think the code will break. It should just skip over any Java files that are not within the `JAVA_SOURCE_DIR`. Meaning, if the `.java` files are outside of `src/main/java`, it will be ignored. There is always possibility of bugs, but I have iterated over directories without problems so far.

Reply:
[Issue #24](https://github.com/jonathan-m-phillips/ASHE_Automated-Software-Hardening-for-Entrypoints/issues/24)